mayastor/src/bdev/nexus/nexus_io.rs

```diff
@@ -1,471 +1,478 @@
-            IoStatus::Pending => 0,
-            IoStatus::Success => 1,
-        }
-    }
+#[derive(Debug)]
+#[repr(C)]
+pub struct NioCtx {
+    in_flight: u8,
+    status: IoStatus,
+    channel: NonNull<spdk_io_channel>,
+    core: u32,
 }
 
-impl From<i8> for IoStatus {
-    fn from(status: i8) -> Self {
-        (status as i32).into()
-    }
-}
-/// NVMe Admin opcode, from nvme_spec.h
-pub mod nvme_admin_opc {
-    // pub const GET_LOG_PAGE: u8 = 0x02;
-    pub const IDENTIFY: u8 = 0x06;
-    // pub const ABORT: u8 = 0x08;
-    // pub const SET_FEATURES: u8 = 0x09;
-    // pub const GET_FEATURES: u8 = 0x0a;
-    // Vendor-specific
-    pub const CREATE_SNAPSHOT: u8 = 0xc0;
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+#[repr(C)]
+enum Disposition {
+    /// All IOs are completed, and the final status of the IO should be set to
+    /// the enum variant
+    Complete(IoStatus),
+    /// IOs are still in flight status of the last IO that failed
+    Flying(IoStatus),
 }
 
-impl Bio {
-    /// obtain tbe Bdev this IO is associated with
-    pub(crate) fn bdev_as_ref(&self) -> Bdev {
-        unsafe { Bdev::from(self.0.as_ref().bdev) }
-    }
-
-    pub(crate) fn io_channel(&self) -> *mut spdk_io_channel {
-        unsafe { spdk_bdev_io_get_io_channel(self.0.as_ptr()) }
-    }
-
-    /// initialize the ctx fields of an spdk_bdev_io
-    pub fn init(&mut self) {
-        self.ctx_as_mut_ref().io_attempts = self.nexus_as_ref().max_io_attempts;
-    }
-
-    /// reset the ctx fields of an spdk_bdev_io to submit or resubmit an IO
-    pub fn reset(&mut self, in_flight: usize) {
-        self.ctx_as_mut_ref().in_flight = in_flight as i8;
-        self.ctx_as_mut_ref().status = IoStatus::Success;
-    }
-
-    /// complete an IO for the nexus. In the IO completion routine in
-    /// `[nexus_bdev]` will set the IoStatus for each IO where success ==
-    /// false.
-    #[inline]
-    pub(crate) fn ok(&mut self) {
-        if cfg!(debug_assertions) {
-            // have a child IO that has failed
-            if self.ctx_as_mut_ref().status != IoStatus::Success {
-                debug!("BIO for nexus {} failed", self.nexus_as_ref().name)
-            }
-            // we are marking the IO done but not all child IOs have returned,
-            // regardless of their state at this point
-            if self.ctx_as_mut_ref().in_flight != 0 {
-                debug!("BIO for nexus marked completed but has outstanding")
-            }
+pub(crate) fn nexus_submit_io(mut io: NexusBio) {
+    if let Err(e) = match io.cmd() {
+        IoType::Read => io.readv(),
+        // these IOs are submitted to all the underlying children
+        IoType::Write | IoType::WriteZeros | IoType::Unmap => io.submit_all(),
+        IoType::Flush => {
+            io.ok();
+            Ok(())
         }
-        unsafe {
-            spdk_bdev_io_complete(self.0.as_ptr(), IoStatus::Success.into())
+        IoType::NvmeAdmin => {
+            io.fail();
```

Comment:
Can we print error message here in addition to returning error, so we can easily see that happening in logs.

Reply:
The match returns the Error variant and if its an error, we print it at line 132.